addons/pos_loyalty/static/src/overrides/models/pos_store.js

```diff
@@ -9,14 +9,298 @@ import { TextInputPopup } from "@point_of_sale/app/utils/input_popups/text_input
 import { Domain, InvalidDomainError } from "@web/core/domain";
 import { ask, makeAwaitable } from "@point_of_sale/app/store/make_awaitable_dialog";
 import { PartnerList } from "@point_of_sale/app/screens/partner_list/partner_list";
+import { Mutex } from "@web/core/utils/concurrency";
+import { effect } from "@web/core/utils/reactive";
+import { batched } from "@web/core/utils/timing";
 
 let nextId = -1;
+const mutex = new Mutex();
+const updateRewardsMutex = new Mutex();
+const pointsForProgramsCountedRules = {};
 
 export function loyaltyIdsGenerator() {
     return nextId--;
 }
 
+function inverted(fn) {
+    return (arg) => !fn(arg);
+}
+
 patch(PosStore.prototype, {
+    async setup() {
+        await super.setup(...arguments);
+
+        this.couponByLineUuidCache = {};
+        this.rewardProductByLineUuidCache = {};
+
+        effect(
+            batched((orders) => {
+                const order = Object.values(orders).find(
+                    (order) => order.uuid === this.selectedOrderUuid
+                );
+
+                if (order) {
+                    this.updateOrder(order);
+                }
+            }),
+            [this.data.records["pos.order"]]
```

Comment:
did you try using the current order directly? So instead of passing all the orders here, we only pass "this.get_order()"?